autotest/ogr/ogr_flatgeobuf.py

```diff
@@ -1133,3 +1133,29 @@ def test_ogr_flatgeobuf_arrow_stream_numpy():
     ds = None
 
     ogr.GetDriverByName("FlatGeobuf").DeleteDataSource("/vsimem/test.fgb")
+
+
+def test_ogr_flatgeobuf_issue_7401():
+    ds = ogr.GetDriverByName("FlatGeobuf").CreateDataSource("/vsimem/test.fgb", options=["SPATIAL_INDEX=NO"])
+    lyr = ds.CreateLayer("test", geom_type=ogr.wkbPoint)
```

Comment:
Ah! So I had not really disabled the spatial index creation... that is a good lead, thanks @rouault.

Reply:
Hmm.. so that fixes the case without spatial index..  but what it causing the corruption with index I'm getting out of ideas. I would expect current code to run, at least to find and be able to parse out the first/single feature due to that m_featureItems is what is written out and it only contains the feature with geometry. Note to self, the case of no geometries with geometry is special and should probably fall back to a non spatially indexed file.

Reply:
I suspect that when index is enabled, the reason is that the file is written in 2 steps:
- one step in a temporary file where features are written in the order of CreateFeature() calls
- a final step where the temporary file is read in a non-sequential way to order the features according to the index. That should be modified to also append at the end the non-spatial features. You'd need to maintain a list of them, likely (offset, size), in CreateFeature()

Reply:
That is more or less what I have in mind but before doing that it doesn't make sense I get with corrupted buffers when the temp file includes the null geoms, it should only mean I still get a partial file (with wrong total feature count). I feel I need to solve that before trying to implement the actual support for null geoms features at the end.

Reply:
try this patch:
```
diff --git a/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp b/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp
index 8106566675..a34dd90098 100644
--- a/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp
+++ b/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp
@@ -636,7 +636,7 @@ bool OGRFlatGeobufLayer::CreateFinalFile()
             return true;
         };
 
-        for (size_t i = 0; i < m_featuresCount; i++)
+        for (size_t i = 0; i < m_featureItems.size(); i++)
         {
             const auto &featureItem = m_featureItems[i];
             const auto featureSize = featureItem.size;
```

Reply:
Of course... 🙈

Reply:
Hmm while clearly a problem and valid logical fix it is not enough to enable it to run without corrupt buffers. I had actually disabled that logic entirely with bUseBatchStrategy = false... Perhaps it's due to that the tree becomes smaller than calculated elsewhere and offsets become all wrong.

Reply:
This is indeed a problem. I use the feature count to determine the size of the tree when reading... I don't see how I can handle this in an un-hacky way without actually adding additional metadata in the header (i.e additional count so that it's known how many are indexed and not up front).

I'm considering making it explicitly forbidden to create spatially indexed flatgeobuf with null geoms. I have difficulties seeing why it is a use case worth supporting.

Reply:
> I'm considering making it explicitly forbidden to create spatially indexed flatgeobuf with null geoms.

That could be reasonable given what the spec currently supports